calendar/api/parent/ext-calendar-provider.js

```diff
@@ -103,11 +104,49 @@ class ExtCalendarProvider extends cal.provider.BaseClass {
           : ["unsubscribe"];
       case "requiresNetwork":
         return !(this.capabilities.requires_network === false);
+      case "itip.transport":
+        if (this.extension.emitter.has("calendar.provider.onSend")) {
+          return this;
+        }
+        break;
     }
 
     return super.getProperty(name);
   }
 
+  scheme = "mailto";
+
+  async sendItems(aRecipients, aItipItem) {
+    let method = aItipItem.responseMethod;
+    let transport = super.getProperty("itip.transport").wrappedJSObject;
+    let { subject, body } = transport._prepareItems(aItipItem);
+    let serializer = Cc["@mozilla.org/calendar/ics-serializer;1"].createInstance(Ci.calIIcsSerializer);
+    let itemList = aItipItem.getItemList();
+    serializer.addItems(itemList, itemList.length);
+    let methodProp = cal.getIcsService().createIcalProperty("METHOD");
+    methodProp.value = method;
+    serializer.addProperty(methodProp);
+    let icsText = serializer.serializeToString();
+    let boundary = Array.from({ length: 24 }, () => Math.floor(Math.random() * 16).toString(16)).join("");
+    let headers = new Map();
+    headers.set("Date", [new Date()]);
+    headers.set("Subject", [subject]);
+    headers.set("To", aRecipients.map(attendee => ({ name: attendee.commonName, email: attendee.id.replace(/^mailto:/, "")})));
+    headers.set("Content-Type", ["multipart/mixed; boundary=\"" + boundary + "\""]);
+    let mimeContent = jsmime.headeremitter.emitStructuredHeaders(headers, { hardMargin: 800 }) + "\r\n";
+    mimeContent += "--" + boundary + "\r\n";
+    mimeContent += "Content-Type: text/plain; charset=UTF-8\r\n\r\n";
+    mimeContent += body + "\r\n";
+    mimeContent += "--" + boundary + "\r\n";
+    mimeContent += "Content-Type: text/calendar; method=" + aItipItem.responseMethod + "; charset=UTF-8\r\nContent-Transfer-Encoding: 8BIT\r\n\r\n";
+    mimeContent += icsText;
+    mimeContent += "--" + boundary + "\r\n";
+    mimeContent += "Content-Type: application/ics; name=\"invite.ics\"\r\nContent-Disposition: attachment; filename=\"invite.ics\"\r\nContent-Transfer-Encoding: 8BIT\r\n\r\n";
+    mimeContent += icsText;
+    mimeContent += "--" + boundary + "--\r\n";
+    await this.extension.emit("calendar.provider.onSend", this, mimeContent, aRecipients.map(attendee => attendee.id.replace(/^mailto:/, "")));
```

Comment:
This assumes that the extension will be using email to send the scheduling message. I think we should instead pass either a js object with the specific details such as responseMethod, a MailExtensions calendar item, etc.

Together with below comment on similarities to webRequest, I think the extension can then either:
* Decide for itself how to send (e.g. maybe make server requests) and indicate it has handled it
* Make some changes to the data and return those, having Thunderbird take care of the send via email
* Leave the message unchanged and have Thunderbird take care of the send via email.

Reply:
I've added a commit to demonstrate that approach as best as I can. I don't know enough about Lightning internals to attempt to change anything other than the list of recipients. The extension is passed useful information about the email, and can return `true` to indicate that it has processed the message, or an object with a `recipients` property to show which recipients it has processed. If the extension does not process any recipients, the default Lightning processing then applies.

Reply:
@kewisch : I can see the value of having the individual properties available. Neil is suggesting to do both, passing the individual properties and the MIME message into the listener, and let the listerer use what it needs.

The problem is that creating the MIME message isn't trivial, as you can see from the code here. I actually think that the whole multipart code should also move to JSMime (but that's for another enhancement request later, to provide these APIs). Worse, it needs JSMime, which is chrome code, so we cannot just copy this code into the extension. Either way, I don't think we should require every calendar provider to re-construct the MIME message, but let Lightning do that consistently for all calendar providers, at least optionally, to ensure consistency and to avoid code duplication.

> Make some changes to the data

While I can see the theoretical value, we don't have actual use cases for that. Additionally, Neil said that the Lightning APIs don't make that easy. Could we leave that to another PR? Given that we don't actually need that, and this here is already progress?

Reply:
> @kewisch : I can see the value of having the individual properties available. Neil is suggesting to do both, passing the individual properties and the MIME message into the listener, and let the listerer use what it needs.

To me this should be up to the extension to do. If the extension needs a mime message they can put it together with a library. If there is no library then we should make jsmine work without xpcom so people can make use of it. jsmine is available without Thunderbird afaik?

If we always provide both formats then we're doing extra conversion where it may not be needed. I don't actually think every calendar provider will need a mime message.

> 
> > Make some changes to the data
> 
> 
> 
> While I can see the theoretical value, we don't have actual use cases for that. Additionally, Neil said that the Lightning APIs don't make that easy. Could we leave that to another PR? Given that we don't actually need that, and this here is already progress?

We can certainly leave it to another pr, maybe add a TODO comment though. I think there would be valid use cases for this.

Reply:
Shipping an entire third party MIME library is not going to work for us. This is inherently the job of the MUA and not of the extension.

Reply:
The code here, in combination with a library, does not seem overly complex to me. If the extension needs a specific format to send that is their job as well.

I could imagine a similar format as the formats specifier where the developer can request specific formats. This would align with the items api where you can select specific formats. That part isn't ideal either, if we could go with one authoritative format that extensions can convert to their liking that would be my preference there as well.

Reply:
> in combination with a library

We will not ship an MIME library with our extension.

> the developer can request specific formats

That's fine with me. Do you want to suggest an API for that?

Reply:
(What I'd like to see, presumably once this becomes part of Thunderbird, is that the code in Lightning that is used to create the MIME message is refactored so that it can be called directly rather than having to mostly cut and paste it like I did here.)

> the developer can request specific formats

I can implement a `returnFormat` option like you have for the other listeners.

Reply:
I did have to shuffle things around a bit so that the option could be accessed but now there are several formats:
- an array of items, in default, ical or jcal format;
- a subject, body and ics text; or
- a MIME message

Reply:
> I did have to shuffle things around a bit so that the option could be accessed

Actually, I find that this new version is more readable, so it's a good change.

Good work!

It's nit-picking about names, but: I find that `onSend` was a better name, because in our case, the `onSend()` function does the actual sending, so that name is appropriate. If the `onSend` is not sending, then Lightning uses SMTP as fallback, so that still makes sense in that case. But I leave the final decree to Philipp.